doc/man7/provider-digest.pod

```diff
@@ -0,0 +1,212 @@
+
+=head1 SYNOPSIS
+
+ #include <openssl/core_numbers.h>
+ #include <openssl/core_names.h>
+
+ /*
+  * None of these are actual functions, but are displayed like this for
+  * the function signatures for functions that are offered as function
+  * pointers in OSSL_DISPATCH arrays.
+  */
+
+ /* Context management */
+ void *OP_digest_newctx(void *provctx);
+ void OP_digest_freectx(void *dctx);
+ void *OP_digest_dupctx(void *dctx);
+
+ /* Digest generation */
+ int OP_digest_init(void *dctx);
+ int OP_digest_update(void *dctx, const unsigned char *in, size_t inl);
+ int OP_digest_final(void *dctx, unsigned char *out, size_t *outl,
+                     size_t outsz);
+ int OP_digest_digest(void *provctx, const unsigned char *in, size_t inl,
+                      unsigned char *out, size_t *outl, size_t outsz);
+
+ /* Digest parameters */
+ size_t OP_digest_size(void);
+ size_t OP_digest_block_size(void);
+ int OP_digest_set_params(void *dctx, const OSSL_PARAM params[]);
+ int OP_digest_get_params(void *dctx, OSSL_PARAM params[]);
+
+=head1 DESCRIPTION
+
+This documentation is primarily aimed at provider authors. See L<provider(7)>
+for further information.
+
+The DIGEST operation enables providers to implement digest algorithms and make
+them available to applications via the API functions L<EVP_DigestInit_ex(3)>,
+L<EVP_DigestUpdate(3)> and L<EVP_DigestFinal(3)> (and other related functions).
+
+All "functions" mentioned here are passed as function pointers between
+F<libcrypto> and the provider in B<OSSL_DISPATCH> arrays via
+B<OSSL_ALGORITHM> arrays that are returned by the provider's
+provider_query_operation() function
+(see L<provider-base(7)/Provider Functions>).
+
+All these "functions" have a corresponding function type definition
+named B<OSSL_{name}_fn>, and a helper function to retrieve the
+function pointer from an B<OSSL_DISPATCH> element named
+B<OSSL_get_{name}>.
+For example, the "function" OP_digest_newctx() has these:
+
+ typedef void *(OSSL_OP_digest_newctx_fn)(void *provctx);
+ static ossl_inline OSSL_OP_digest_newctx_fn
+     OSSL_get_OP_digest_newctx(const OSSL_DISPATCH *opf);
+
+B<OSSL_DISPATCH> arrays are indexed by numbers that are provided as
+macros in L<openssl-core_numbers.h(7)>, as follows:
+
+ OP_digest_newctx        OSSL_FUNC_DIGEST_NEWCTX
+ OP_digest_freectx       OSSL_FUNC_DIGEST_FREECTX
+ OP_digest_dupctx        OSSL_FUNC_DIGEST_DUPCTX
+
+ OP_digest_init          OSSL_FUNC_DIGEST_INIT
+ OP_digest_update        OSSL_FUNC_DIGEST_UPDATE
+ OP_digest_final         OSSL_FUNC_DIGEST_FINAL
+ OP_digest_digest        OSSL_FUNC_DIGEST_DIGEST
+
+ OP_digest_size          OSSL_FUNC_DIGEST_SIZE
+ OP_digest_block_size    OSSL_FUNC_DIGEST_BLOCK_SIZE
+ OP_digest_set_params    OSSL_FUNC_DIGEST_SET_PARAMS
+ OP_digest_get_params    SSL_FUNC_DIGEST_GET_PARAMS
+
+A digest algorithm implementation may not implement all of these functions.
+In order to be useable all or none of OP_digest_newctx, OP_digest_freectx,
```

Comment:
All or none?  So it's usable if only the optional functions are implemented?
...
Oh, right, if `OSSL_FUNC_DIGEST_DIGEST` is implemented, then it's still useful.

Reply:
Yes - exactly.